### PR TITLE
[SBOM] Replace BadgerDB cache by a custom LRU cache with a cleaning mechanism for Trivy

### PR DESCRIPTION
**Pull request by AliDatadog**
### What does this PR do?
Replaces `BadgerDB` cache by a new custom LRU cache for `Trivy` with a local storage backed by `BoltDB` and a clean up mechanism to remove unused entries. It has mechanisms to limit the size in memory and on disk.
The clean up mechanism required to store `ArtifactIDs` and `BlobIDs` in the `workloadmeta.SBOM` so that they can be removed from the cache. 

It adds config options to parametrize the cache:
    - `container_image_collection.sbom.use_custom_cache`: Enables the custom cache.
    - `container_image_collection.sbom.custom_cache_max_disk_size`: Maximum total size of the cached objects. **It is not equivalent to disk usage. If `BoltDB` uses extra storage, it will not limited by this config option.**
    - `container_image_collection.sbom.custom_cache_max_cache_entries`: Maximum number of cached objects.
    - `container_image_collection.sbom.cache_clean_interval `: Duration interval of garbage collection.

It also adds the following metrics to monitor the custom cache:
- `sbom_cache_disk_size`: Total size, in bytes, of the database as reported by `BoltDB`.
- `sbom_cached_object_size`: Total size, in bytes, of cached SBOM objects on disk. Limited by `container_image_collection.sbom.custom_cache_max_disk_size`
- `sbom_cache_hits_total`: Total number of cache hits
- `sbom_cache_misses_total`: Total number of cache misses
- `sbom_cache_evicts_total`: Total number of cache evicts
 

### Motivation

The default Trivy cache based on `Boltdb` doesn’t work for us, because it doesn’t have any cleanup mechanism. It is ok for a CLI based approach, but for a long running process like the Datadog-Agent it can trigger an high memory and disc usage.

We tried to implement a custom cache based on badgerDB that provide a TTL mechanism, but after some tests we discovered that the DB doesn't fit other criterias that we have:

- BadgerDB doesn’t offer an easy way to update TTL of a specific entry

- BadgerDB is optimized for large dataset (millions of entries). But it is not our case.

- BadgerDB allocate a large disk space event if it doesn’t use it.

### Additional Notes
If during a Trivy scan, a cache entry required by the scan is deleted then the scan will be aborted. 

We can parametrize the maximum number of entries and the maximum size of cached objects. 

### Possible Drawbacks / Trade-offs
- Memory and disk footprint to measure.

### Describe how to test/QA your changes
The PR is divided into multiple commits that can be reviewed one by one.

Verify that the unit tests are passing. 

Deploy the agent on an environment with `containerd` and these settings:

`DD_CONTAINER_IMAGE_COLLECTION_METADATA_ENABLED=true`
`DD_CONTAINER_IMAGE_COLLECTION_SBOM_ENABLED=true`
`DD_CONTAINER_IMAGE_COLLECTION_SBOM_USE_CUSTOM_CACHE=true`
`DD_CONTAINER_IMAGE_COLLECTION_SBOM_CACHE_CLEAN_INTERVAL=180`

- Deploy the agent and pull an image. Check the SBOM generation time with agent workload-list --verbose. In my case, the `datadog/agent:latest` images takes a few seconds. 
- Now delete the image, pull it again, and check the generation time again. It should be faster because of the cache. 
- Delete the image again and wait for the garbage collector which should run every 180s with the settings above.
- Pull the image again, it should take a few seconds.

To test the max disk size and max cache entries, the unit tests should be enough. Otherwise you can add log lines to `MissingBlobs`, set low values to
`DD_CONTAINER_IMAGE_COLLECTION_SBOM_CUSTOM_CACHE_MAX_CACHE_ENTRIES` and `DD_CONTAINER_IMAGE_COLLECTION_SBOM_CUSTOM_CACHE_MAX_DISK_SIZE` and pull many images. Be careful, if the values are too low, `GetArtifact` and `GetBlob` will return an error, making the scan fail. You can also monitor the metrics listed above. Finally, it might also be possible to inspect the db with the `BoltDB` CLI.

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature. 
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

**Comment by AliDatadog** [comment]
This PR will first be merged on a custom 7.44.x branch and tested on staging. Then it will eventually be merged on main.

**Review by pgimalac — APPROVED** [review]
LGTM for config file.

File: releasenotes/notes/add-custom-local-cache-sbom-d64e46b509e40161.yaml [hunk 1]
```diff
@@ -0,0 +1,14 @@
+enhancements:
+  - |
+    * Add new custom lru cache for Trivy backed by BoltDB and parametrized by:
```
Comment (maycmlee) [inline releasenotes/notes/add-custom-local-cache-sbom-d64e46b509e40161.yaml:3]:
```suggestion
    * Add new custom LRU cache for Trivy backed by BoltDB and parametrized by:
```

**Review by maycmlee — APPROVED** [review]
small edit, but approving

File: pkg/util/trivy/cache.go [hunk 2]
```diff
@@ -186,3 +215,292 @@ func (c *TrivyCache) GetArtifact(id string) (types.ArtifactInfo, error) {
...
+	for c.currentCachedObjectTotalSize > target {
+		err := c.removeOldest()
+		if err != nil {
+			return err
+		}
+		if prev == c.currentCachedObjectTotalSize {
+			// if c.currentCachedObjectTotalSize is not updated by removeOldest then an item is stored in the lrucache without being stored in the local storage
+			return fmt.Errorf("cache and db are out of sync")
+		}
+	}
```
Comment (L3n41c) [inline pkg/util/trivy/cache.go]: In order to check that every single call to `c.removeOldest()` leads to a modification of `c.currentCachedObjectTotalSize`, the line

```go
prev := c.currentCachedObjectTotalSize
```

needs to be moved inside the `for c.currentCachedObjectTotalSize > target {` loop.

Comment (AliDatadog) [reply pkg/util/trivy/cache.go]: Good catch ! I must have erased the line that updates prev by mistake. I wonder if we should just return an error here ? If the cache and db are out of sync, I think most operations will fail.

File: pkg/util/trivy/cache.go [hunk 3]
```diff
@@ -186,3 +215,292 @@ func (c *TrivyCache) GetArtifact(id string) (types.ArtifactInfo, error) {
...
+	defer c.mutex.Unlock()
+	return c.removeOldest()
+}
+
+// reduceSize reduces the size of the cache to the target size by evicting the oldest items.
+func (c *PersistentCache) reduceSize(target int) error {
+	if c.currentCachedObjectTotalSize <= target {
+		return nil
+	}
+
```
[the remark this answers is not in the document]
Comment (AliDatadog) [reply pkg/util/trivy/cache.go]: It should be fixed by the last commit.

File: pkg/workloadmeta/types.go [hunk 4]
```diff
@@ -654,6 +654,8 @@ type SBOM struct {
 	CycloneDXBOM       *cyclonedx.BOM
 	GenerationTime     time.Time
 	GenerationDuration time.Duration
+	ArtifactID         string   // Required for cache cleaning
+	BlobIDs            []string // Required for cache cleaning
```
[the remark this answers is not in the document]
Comment (AliDatadog) [reply pkg/workloadmeta/types.go]: Indeed, I also think we do not need to expose them.

Instead, I propose to store the mapping `ImageID => ArtifactID/BlobID" directly in the cache cleaner. It is done in my last commit.

File: pkg/config/config.go [hunk 5]
```diff
@@ -1115,7 +1115,10 @@ func InitConfig(config Config) {
 	config.BindEnvAndSetDefault("container_image_collection.sbom.analyzers", []string{"os"})
 	config.BindEnvAndSetDefault("container_image_collection.sbom.cache_directory", "")
 	config.BindEnvAndSetDefault("container_image_collection.sbom.clear_cache_on_exit", false)
-	config.BindEnvAndSetDefault("container_image_collection.sbom.cache_ttl", 60*60) // Integer seconds. Only used with Badger
+	config.BindEnvAndSetDefault("container_image_collection.sbom.use_custom_cache", false)
+	config.BindEnvAndSetDefault("container_image_collection.sbom.custom_cache_max_disk_size", 1000*1000*10) // max disk space used by cached objects. Not equal to max disk usage
+	config.BindEnvAndSetDefault("container_image_collection.sbom.custom_cache_max_cache_entries", 1000)     // keys stored in memory
+	config.BindEnvAndSetDefault("container_image_collection.sbom.cache_clean_interval", 60*30)              // only useful for custom cache
```
Comment (L3n41c) [inline pkg/config/config.go:1121]: This parameter is used as a parameter of `GetDuration(…)`, so the best would be to provide a duration string.

```suggestion
	config.BindEnvAndSetDefault("container_image_collection.sbom.cache_clean_interval", "30m")              // only useful for custom cache
```

File: pkg/workloadmeta/collectors/internal/containerd/image_sbom_trivy.go [hunk 6]
```diff
@@ -98,15 +98,23 @@ func (c *collector) startSBOMCollection(ctx context.Context) error {
 				log.Warnf("Unable to close trivy client: %v", err)
 			}
 		}()
+		cleanTicker := time.NewTicker(config.Datadog.GetDuration("container_image_collection.sbom.cache_clean_interval") * time.Second)
```
Comment (L3n41c) [inline pkg/workloadmeta/collectors/internal/containerd/image_sbom_trivy.go:101]: `GetDuration(…)` already returns a `time.Duration`. So, we shouldn’t multiply its return value by `time.Second` as we would with an `int` in order to turn it into a `time.Duration`.

```suggestion
		cleanTicker := time.NewTicker(config.Datadog.GetDuration("container_image_collection.sbom.cache_clean_interval"))
```

**Pull request merged**